Feature repository stars

Vasilijez opened the pull request:
Closes #38.

All mentioned requests in #38 have been implemented.

Some comments:
- On the repository page, if the user meets the conditions, they will be able to star the repository.
- User's starred repositories are available at `http://localhost:5173/u/{username}/starred`.

Vasilijez committed:
feat: Create model

Vasilijez committed:
feat: Add `get_starred_repositories_of_user` API

Vasilijez committed:
feat: Add `toggle_repo_star` API

Vasilijez committed:
feat: Add access control for repository star toggle

Vasilijez committed:
feat: Add 'stars' attribute to 'Repository'

Vasilijez committed:
feat: Support star toggle button with tooltip

Vasilijez committed:
chore: Remove accidentally committed `logger` line

Vasilijez committed:
feat: Enhance toast message with more detailed information

Vasilijez committed:
feat: Add `RepoStarred` page with `RepoPreview` component

Vasilijez committed:
feat: Use `RepoPreview` component in `RepoOfUser` page

Vasilijez committed:
fix: Add utility class for vertical alignment

Vasilijez committed:
fix: Improve comments for better comprehension

Vasilijez committed:
tests: Add unit tests for `toggle_repo_star`

Vasilijez committed:
fix: Simplify model

Vasilijez committed:
feat: Add `has_star_access`

Previously, not all cases were covered. Due to complexity, creating a separate function in the access control service makes sense.

Vasilijez committed:
test: Write integration test for `toggle_repo_star`

Vasilijez committed:
refactor: Improve consistency

Vasilijez committed:
fix: Prevent loading from masking error messages

Vasilijez committed:
feat: Return an error message instead of an empty list

Vasilijez committed:
test: Write unit test for `has_star_access`

Vasilijez committed:
test: Write integration test for `get_starred_repositories_of_user`

Vasilijez committed:
test: Write and expand the existing integration test for `get_repo_by_canonical_name`

I have written the test without altering the existing implementation. I have not written all cases, as many are already covered in other integration tests.

Vasilijez committed:
Merge branch 'develop' into feature-repository-stars

Vasilijez committed:
fix: Resolve conflict of `repo.last_updated`

There was a conflict after merging `develop` into this branch. The incoming updated property (`repo.last_updated`) has been moved to the newly created `RepoPreview.tsx` component.

magley commented on client/src/pages/RepoPage.tsx:
```diff
@@ -31,6 +33,24 @@ export const RepositoryPage: React.FC = () => {
+                ...repo,
+                ...res.data,
+            });
+            addToast(`${repo.name} is ${repo.starred ? "starred" : "unstarred"}.`, ToastType.success)
```
The toast message seems to be flipped (it says "starred" when it should be "unstarred" and vice versa):
![image](https://github.com/user-attachments/assets/9a9db06b-180d-4a6e-859b-8bd06b11fe44)

Vasilijez replied:
Nice catch! I have used the `repo` state, but React doesn't reflect updated changes in the `.then()` callback since it's interpreted in the same render cycle. The solution is to use `res.data`. 

Commit: [463dfb2](https://github.com/magley/mocker-hub/pull/39/commits/463dfb20a691a6627f22086de2db87dfc42f0896)

magley commented on client/src/pages/RepoStarred.tsx:
```diff
@@ -0,0 +1,69 @@
+                fullResult?.repos.map((repo) => (
+                    <RepoPreview key={repo.id} repo={repo} orgNames={orgNames} />
+                ))
+            }
```
Could you add some text for when the user hasn't starred any repositories, so that the page isn't empty?

Vasilijez replied:
Commit: [c1225b0](https://github.com/magley/mocker-hub/pull/39/commits/c1225b0224ce7d4f5cb0d8f98c40fe4f8d807d97)

magley commented on server/app/api/repo/repo_controller.py:
```diff
@@ -70,8 +72,10 @@ def get_repo_by_canonical_name(
     result["org_name"] = None if (repo.organization is None) else repo.organization.name
     result["can_update"] = access_control_service.has_write_access(user_id, repo.id)
+    result["can_star"] = access_control_service.has_star_access(user_id, repo.id)
+    result["starred"] = any(star.repository.id == repo.id for star in user.stars) if (result["can_star"]) else False
```
This code:
`star.repository.id == repo.id for star in user.stars`
also appears in `RepoService::toggle_repo_star`. Extract it into a service method `is_repo_starred_by(repo, user)` or something like that.

Vasilijez replied:
Commit: [61ad8fe](https://github.com/magley/mocker-hub/pull/39/commits/61ad8feb048d557e04cb1cb739fff4ba7c54325f)

magley commented on server/app/api/repo/repo_controller.py:
```diff
@@ -106,3 +110,39 @@ def update_repo_visibility_by_id(
+
+    org_names = org_service.find_org_names_by_ids([r.organization_id for r in repos if r.organization_id is not None])
+
+    return StarredReposOfUserDTO(user_id=user_id, user_name=user.username, repos=repos, organization_names=org_names)
```
The code in this route handler is very similar to the one in the `get_repositories_of_user()` route handler:
```

def get_repos_of_user():
    user = get_user()
    repos = get_repos()
    orgs = get_orgs()
    
    repos = model_to_dto(repos)
    return 

def get_starred_repos_of_user():
    user = get_user()
    starred_repos = get_starred_repos ()
    orgs = get_orgs()
    
    starred_repos = model_to_dto(starred_repos )
    return
```

So I propose you deduplicate some of the code:

- Add a `def repo_model_to_dto(r: Repository) -> RepositoryDTO` utility function in the controller which calls the `model_dump`
- Add a `def get_org_names_from_repos(repos: Repository[]) -> str[]` utility function in the controller that internally calls `org_service.find_org_names_by_ids`. The parameter is either `Repository[]` or `RepositoryDTO[]`. I think either will work.
- Add a `get_starred_repositories_of_user(u: User) -> Repository[]` service method

I should've done this myself when I wrote `get_repositories_of_user`. Controller methods should be as simple as possible.

This is all me refactoring in my head, if you notice that something won't work, feel free to adjust. Similarly, if you think some of my suggestions result in code that is too "rigid", you can try a different approach, or leave it as-is.

Vasilijez replied:
Commit: [61ad8fe](https://github.com/magley/mocker-hub/pull/39/commits/61ad8feb048d557e04cb1cb739fff4ba7c54325f)

magley commented on server/app/api/repo/repo_dto.py:
```diff
@@ -29,11 +29,22 @@ class ReposOfUserDTO(BaseModel):
+    user_id: int
+    user_name: str
+    repos: List[RepositoryDTO]
+    organization_names: Dict[int, str]
```
`StarredReposOfUserDTO` and `ReposOfUserDTO` share the same fields. Since they also serve the same purpose (almost), maybe `StarredReposOfUserDTO` should inherit from `ReposOfUserDTO`. Are there any scenarios in the future we'd want to keep them separate? Maybe `StarredReposOfUserDTO` doesn't even need to exist and `ReposOfUserDTO` can be used for starred repos as well. I'll let you decide.

Vasilijez replied:
I doubted that for so long, but after implementing functionalities, I think there are no scenarios where `StarredReposOfUserDTO` will be different from `ReposOfUserDTO`.

Commit: [61ad8fe](https://github.com/magley/mocker-hub/pull/39/commits/4178ebb04de70fbb5e4f09c747ca90a5ff5f331f)

magley commented on server/app/api/repo/repo_service.py:
```diff
@@ -108,5 +108,31 @@ def update_repo_by_id(self, repo_id: int, dto: RepositoryDescUpdateDTO | Reposit
+
+        repo = self.repo_repo.unstar_repo(repo, user)
+        if repo is None:
+            raise NotFoundException(RepositoryStar, f"user_id, repo_id: {user_id}, {repo_id}")
```
Can [131-132] ever happen? `repo.unstar_repo` [130] is called just before this check, and that method returns None only if the repo isn't already starred. If the repo isn't starred, then it will enter the if block in [125-128].
I understand that you wanted the repository to be fail-safe, which is why you check if the repo is starred before unstarring it, but in case the repo **isn't** starred and you try to unstar it, it should return the repo itself (and not `None`).

In other words, [131-132] isn't necessary, and in `repo_repo::unstar_repo()`, if `repo_star is None` then return the `repo`. Would that work, though? Maybe I'm missing something.

Vasilijez replied:
I was thinking about some concurrent situations, like if one user sends two requests one after another. However, I will remove this additional check. Transaction handling requires deeper examination, and we are not required to pay attention to conflict situations in this student project.

Commit: [f4cc379](https://github.com/magley/mocker-hub/pull/39/commits/f4cc3795e8a321c567187812e589d9020956fe39)

magley reviewed (changes requested):
Works nicely, most of my comments are either minor web-related things or some suggestions for refactoring the code. I didn't look at the tests, they seem exhaustive and if they pass then that's alright.

I like that you kept all the commits isolated from each other.

Vasilijez committed:
refactor: Use `ReposOfUserDTO` instead of `StarredReposOfUserDTO`

For now, it seems that `StarredReposOfUserDTO` won't have any additional attributes in the future, so switching to `ReposOfUserDTO` for deduplication.

Vasilijez committed:
fix: Use appropriate DTO: `ToggleStarRepoDTO`

Vasilijez committed:
fix: Resolve wrongly flipped `starred`, `unstarred` message

Vasilijez committed:
feat: Show message for no repositories

Vasilijez committed:
refactor: Deduplicate code in controllers

Followed the proposed changes by @magley and refactored the following controllers:  `get_repositories_of_user`, `get_repo_by_canonical_name` and `get_starred_repositories_of_user`.

Vasilijez committed:
fix: Remove potentially dead code statement

Vasilijez committed:
test: Add unit tests after refactoring `repo_controller.py`
